**Alunos: Jefferson Costa, Sávio Berdine, Amanda Lasserre, Kevin Andrews, Nicholas Henrique**


Importação das bibliotecas necessárias:


In [1]:
import pandas as pd 
from sklearn import preprocessing
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report

In [2]:
#importando os datasets
train = pd.read_csv("train.csv")
test = pd.read_csv("test.csv")
submission = pd.read_csv("gender_submission.csv")

In [3]:
test

PassengerId  Pclass                                          Name  \
0            892       3                              Kelly, Mr. James   
1            893       3              Wilkes, Mrs. James (Ellen Needs)   
2            894       2                     Myles, Mr. Thomas Francis   
3            895       3                              Wirz, Mr. Albert   
4            896       3  Hirvonen, Mrs. Alexander (Helga E Lindqvist)   
..           ...     ...                                           ...   
413         1305       3                            Spector, Mr. Woolf   
414         1306       1                  Oliva y Ocana, Dona. Fermina   
415         1307       3                  Saether, Mr. Simon Sivertsen   
416         1308       3                           Ware, Mr. Frederick   
417         1309       3                      Peter, Master. Michael J   

        Sex   Age  SibSp  Parch              Ticket      Fare Cabin Embarked  
0      male  34.5      0      0              330911    7.8292   NaN        Q  
1    female  47.0      1      0              363272    7.0000   NaN        S  
2      male  62.0      0      0              240276    9.6875   NaN        Q  
3      male  27.0      0      0              315154    8.6625   NaN        S  
4    female  22.0      1      1             3101298   12.2875   NaN        S  
..      ...   ...    ...    ...                 ...       ...   ...      ...  
413    male   NaN      0      0           A.5. 3236    8.0500   NaN        S  
414  female  39.0      0      0            PC 17758  108.9000  C105        C  
415    male  38.5      0      0  SOTON/O.Q. 3101262    7.2500   NaN        S  
416    male   NaN      0      0              359309    8.0500   NaN        S  
417    male   NaN      1      1                2668   22.3583   NaN        C  

[418 rows x 11 columns]

In [4]:
#selecionando as features relevantes pra treino e pra teste
features = ["Pclass","Sex","Age","SibSp","Parch"]
x_test = pd.DataFrame(test.loc[:,features].values)
x_test

0       1     2  3  4
0    3    male  34.5  0  0
1    3  female  47.0  1  0
2    2    male  62.0  0  0
3    3    male  27.0  0  0
4    3  female  22.0  1  1
..  ..     ...   ... .. ..
413  3    male   NaN  0  0
414  1  female  39.0  0  0
415  3    male  38.5  0  0
416  3    male   NaN  0  0
417  3    male   NaN  1  1

[418 rows x 5 columns]

In [5]:
#selecionando a coluna que importa para nossa análise
y_test = submission.loc[:,"Survived"].values

#dividindo o conjunto para o treino entre features e resultados desejados
x_train = pd.DataFrame(train.loc[:,features].values)
y_train = train.loc[:,"Survived"].values

In [6]:
#Transformando as strings em valores númericos
converte = preprocessing.LabelEncoder()
x_train = x_train.apply(converte.fit_transform)
x_test = x_test.apply(converte.fit_transform)

In [7]:
#criando uma instância do classificador
classifier = DecisionTreeClassifier(max_depth=5)

#ajustando o classificador
classifier.fit(x_train, y_train)

DecisionTreeClassifier(max_depth=5)

In [8]:
y_pred = classifier.predict(x_test)
accuracy_score(y_test, y_pred)

0.8492822966507177

In [9]:
report = classification_report(y_test, y_pred)
print(report)

              precision    recall  f1-score   support

           0       0.84      0.94      0.89       266
           1       0.87      0.68      0.77       152

    accuracy                           0.85       418
   macro avg       0.86      0.81      0.83       418
weighted avg       0.85      0.85      0.84       418



**Testando instâncias do classificador para árvores com profundidades diferentes**

Uma instância com profundidade igual a 2:

In [10]:
#criando uma instância do classificador
classifier = DecisionTreeClassifier(max_depth=2)

#ajustando o classificador
classifier.fit(x_train, y_train)

DecisionTreeClassifier(max_depth=2)

In [11]:
y_pred = classifier.predict(x_test)
report = classification_report(y_test, y_pred)
print(report)

              precision    recall  f1-score   support

           0       0.78      0.97      0.86       266
           1       0.90      0.53      0.66       152

    accuracy                           0.81       418
   macro avg       0.84      0.75      0.76       418
weighted avg       0.82      0.81      0.79       418



Uma instância com profundidade igual a 4:

In [12]:
#criando uma instância do classificador
classifier = DecisionTreeClassifier(max_depth=4)

#ajustando o classificador
classifier.fit(x_train, y_train)

DecisionTreeClassifier(max_depth=4)

In [13]:
y_pred = classifier.predict(x_test)
report = classification_report(y_test, y_pred)
print(report)

              precision    recall  f1-score   support

           0       0.99      0.97      0.98       266
           1       0.96      0.98      0.97       152

    accuracy                           0.98       418
   macro avg       0.97      0.98      0.97       418
weighted avg       0.98      0.98      0.98       418



Uma instância com profundidade igual a 6:

In [14]:
#criando uma instância do classificador
classifier = DecisionTreeClassifier(max_depth=6)

#ajustando o classificador
classifier.fit(x_train, y_train)

DecisionTreeClassifier(max_depth=6)

In [15]:
y_pred = classifier.predict(x_test)
report = classification_report(y_test, y_pred)
print(report)

              precision    recall  f1-score   support

           0       0.82      0.95      0.88       266
           1       0.89      0.64      0.74       152

    accuracy                           0.84       418
   macro avg       0.86      0.80      0.81       418
weighted avg       0.85      0.84      0.83       418



Um instância com a profundidade igual a 8:

In [16]:
#criando uma instância do classificador
classifier = DecisionTreeClassifier(max_depth=8)

#ajustando o classificador
classifier.fit(x_train, y_train)

DecisionTreeClassifier(max_depth=8)

In [17]:
y_pred = classifier.predict(x_test)
report = classification_report(y_test, y_pred)
print(report)

              precision    recall  f1-score   support

           0       0.79      0.70      0.74       266
           1       0.56      0.68      0.61       152

    accuracy                           0.69       418
   macro avg       0.68      0.69      0.68       418
weighted avg       0.71      0.69      0.70       418



O teste foi realizado para 4 profundidades diferentes, em relação à primeira profundidade, que foi 2, o teste com produnfidadde 4 e 6 obtiveram melhores resultados na *accuracy*, porém houve uma piora na *precision* para o caso de *survivers*  tanto com a profundidade 6 e 8, quando comparada ap a 2. Além da *precision*, houve pioara também no *recall* e *accuracy* no caso da profundide 8, sendo esta a profundidade que nos forneceu os piores resultados. Por outro lado, o melhor resultado foi obtido tendo uma árvore com profundidade 4, alcançando mais de 96% em todas as métricas. 